Configure direct chain

abailly-iohk opened the pull request:
Completed configuration of direct chain within Hydra node, tests still failing because we are missing carol's key, and probably more issues down the road. The bench are basically not working anymore.

abailly committed:
Add options for passing node socket and signing key

abailly committed:
Can run a Hydra node configured with Direct chain

ETE test still fails to start Hydra Nodes because key for carol is missing.

ghost commented:
To generate a key:
```
Aeson.encodeFile "carol.sk" =<< pure . Api.serialiseToTextEnvelope (Just "Payment Signing Key") =<< Api.generateSigningKey Api.AsPaymentKey
```

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -106,9 +110,25 @@ networkMagicParser =
           <> metavar "MAGIC"
           <> value 42
-          <> help "Network magic for the target network."
+          <> help "Network magic for the target network (default: 42)."
```
There is `showDefault` IIRC

KtorZ commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -106,9 +110,25 @@ networkMagicParser =
+pathToWalletKeyParser :: Parser FilePath
+pathToWalletKeyParser =
+  strOption
+    ( long "wallet-key-file"
```
May I suggest: `wallet-signing-key`

- "file" is already hinted by the metavar
- there are, at least in the context of the wallet, two keys, so it's good to say which one.

KtorZ committed:
Include cardano verification keys to the command-line options

  Also, renamed '--party', '--me' and other options to make it more consistent. Still need to be modified in other interfaces / references such as docker-compose.

ch1bo committed:
Change withHydraNode to also take cardano vkeys as names

ch1bo committed:
Add cardano credentials for carol in local-cluster

abailly committed:
Read cardano verification keys when starting Hydra node

KtorZ committed:
Add partial JSON instance for DirectChainLog

  Currently leveraging 'Show' instances for types that lacks ToJSON intances or for which writing it would be more involved.

ch1bo committed:
Prefix own cardano vkey to cardano keys

ch1bo committed:
Increase active slot coeff to 100% and increase timeouts

abailly committed:
Add initial funds for carol

KtorZ committed:
Pass full paths for cardano keys to withHydraNode

  Instead of relying on signingKeyPathFor in the function's body, to make the function re-usable by the benchmarking code.

KtorZ committed:
Use BFT node in benchmark

  Note that we still need to generate the keys and the proper genesis file for this to work. Right now, keys follow a filepath convention.

abailly committed:
Use cardano-api ToJSON instances in ledger

abailly committed:
Fix compilation errors following options change

KtorZ committed:
Re-introduce default value of --hydra-signing-key (previously --me)

  OptionsSpec rely on this for conciseness.

abailly committed:
Revert "Use cardano-api ToJSON instances in ledger"

This reverts commit abdc93a1954ccfe3f73116808cd9f40f99a4f40c.

abailly committed:
Solve issue with overlapping instances in benchmark code

abailly committed:
Make HydraLog API conformance test pending

We need an Arbitrary instance for DirectChainLog which we don't
currently have.

KtorZ committed:
Move watchUtxoUntil to 'Direct.Wallet'

  This function was in test code, but it is actually relevant for broader use-cases and isn't test-specific.

KtorZ committed:
Introduce new callback / interface for mock-server to wait for next block

  This, in order to allow test code to wait without using unreliable thread delays (even though the current implementation uses thread delays). Another way to approach this would be to watch wallets' UTXOs until some condition is satisfied, but this would require to 'leak' the wallet through the DirectChain component, which we are currently reluctant to.

KtorZ committed:
Implement 'waitForBlock' in the MockServer.

  Doesn't work as we expect, because the 'truth' lies on the client, not on the server. We have to wait until the client has processed a block.

KtorZ committed:
Revert "Implement 'waitForBlock' in the MockServer."

This reverts commit e1dd6883c9705c81302c5a67b64e925e6c750baf.

KtorZ committed:
Revert "Introduce new callback / interface for mock-server to wait for next block"

This reverts commit cfc11142c47dc3b7ab44c7335d7403850b8a14af.

KtorZ committed:
Add simple 'retrying' helper to retry 'postTx' calls.

  This unfortunately doesn't work, because postTx really is only putting a transaction in a queue. What fails is our tx-submission client which is run asynchronously.

KtorZ committed:
Move ValidatedTx construction to postTx' callback instead of tx submit server

  The benefits are plural:

  (a) It gives us a way to catch errors coming from the transaction construction synchronously withint the client code.
  (b) In case of failure, the server is not crashing but, the client -- since the server is running in an async thread and may be connected to multiple clients, this is more resilient.
  (c) It makes the code a lot easier actually; both for the submission and for the server handler which is now quite dump (pop things out of the queue, and submit them as is).

KtorZ committed:
Define Arbitrary instances for direct chain and tiny wallet logs.

KtorZ committed:
Avoid using too large generators for logs

  Some logs contains list of transactions, so generating a list of logs is accidentally quadratic in the size of the generator and, unnecessary to get already good property coverage.

KtorZ committed:
Re-enable logging spec now that arbitrary instances are available.

KtorZ committed:
Replace withCluster with single BFT node

  This is 'temporary' to avoid rollback in tests. Ideally, we need to handle rollback properly.

KtorZ committed:
Factor-in 'waitForSocket' in 'withBFTNode'

  This was a necessary step to call immediately after entering the callback for 'withBFTNode', so it makes more sense to do it within the setup inside 'withBFTNode'.

KtorZ committed:
Remove hadolint from github workflow

  Hadolint is really useful to run when we change the Dockerfile, which happens very seldom; it is quite flaky in the end and often makes the CI fails without much added value.
